q/tree.go

```diff
@@ -0,0 +1,89 @@
+package q
+
+import (
+	"go/token"
+	"reflect"
+)
+
+// A Criteria is used to test against a record to see if it matches.
+// It can be a combination of multiple other criterias.
+type Criteria interface {
+	// Exec is used to test the criteria against a record
+	Exec(interface{}) bool
+	exec(*reflect.Value) bool
+}
+
+type cmp struct {
+	field string
+	value interface{}
+	token token.Token
+}
+
+func (c *cmp) Exec(i interface{}) bool {
+	v := reflect.Indirect(reflect.ValueOf(i))
+	return c.exec(&v)
+}
+
+func (c *cmp) exec(v *reflect.Value) bool {
+	field := v.FieldByName(c.field)
+	return compare(field.Interface(), c.value, c.token)
+}
+
+type or struct {
+	children []Criteria
+}
+
+func (c *or) Exec(i interface{}) bool {
+	v := reflect.Indirect(reflect.ValueOf(i))
+	return c.exec(&v)
+}
+
+func (c *or) exec(v *reflect.Value) bool {
+	for _, criteria := range c.children {
+		if criteria.exec(v) {
+			return true
+		}
+	}
+
+	return false
+}
+
+type and struct {
+	children []Criteria
+}
+
+func (c *and) Exec(i interface{}) bool {
```

Comment:
Yes you're right it'is confusing, i'll rename it.

`Exec` takes an interface then creates a `reflect.Value` from it. While traversing the tree of criterias, if we call the `Exec` function of each criteria we would need to create a `reflect.Value` everytime.

`ExecValue` is here to directly pass the `*reflect.Value`. It is an optimisation but i admit it can be quite confusing.
I will remove it from the interface for now, or unexport it. But i am not sure if an unexported method in an interface prevents the interface from behing implemented outside the package.
I will try